core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -7204,6 +7227,146 @@ protected SqlNode expandDynamicStar(SqlIdentifier id, SqlIdentifier fqId) {
+   * to which the alias evaluates (if there are multiple), and the index into the select list that
+   * said expression occurs at. The information is arranged (numOccurrences, (expr, expr_idx))
+   */
+  protected static Pair<Integer, Pair<@Nullable SqlNode, Integer>> findAliasesInSelect(
+      String name, SqlNameMatcher nameMatcher, SqlSelect select, Integer maxIdx) {
+    SqlNode expr = null;
+    int numOccurrencesOfAlias = 0;
+    int idxOfOccurrence = 0;
+
+    for (int i = 0; i < maxIdx; i++) {
+      SqlNode s = select.getSelectList().get(i);
+      final String alias = SqlValidatorUtil.getAlias(s, -1);
+      if (alias != null && nameMatcher.matches(alias, name)) {
+        expr = s;
+        idxOfOccurrence = i;
+        numOccurrencesOfAlias++;
+      }
+    }
+    return new Pair<>(numOccurrencesOfAlias, new Pair<>(expr, idxOfOccurrence));
+  }
+
+  /**
+   * Shuttle which walks over an expression replacing usage of alias with underlying expression.
+   */
+  static class ExtendedAliasExpander extends Expander {
+    SqlSelect aliasSourceSelect;
+
+    // When recursively checking for aliases, we often encounter situations where we need to check
+    // a subset of the select items for an alias. IE:
+    //    SELECT A, B, C from ...
+    // If we're trying to find aliases for "B" we should only check every select statement that
+    // precedes B. IE, the following statement should fail since the "x" alias
+    // does not precede the select item.:
+    //    SELECT A, x, C as x from ...
+    // but the following statement will work:
+    //    SELECT C as x, x, A from ...
+    //
+    // Note that we must also handle the following case (which is allowed in SF):
+    //    SELECT A as x, x as x from ...
+
+    // This argument instructs the expander to only check elements in the select list
+    // up to but not including this index.
+    Integer maxIdx;
+
+    ExtendedAliasExpander(SqlValidatorImpl validator, SqlValidatorScope scope,
+        SqlSelect select) {
+      // This an alias in the WHERE/ON clause, since we need to check all elements
+      // of the select list for aliases
+      super(validator, scope);
+      this.aliasSourceSelect = select;
+      this.maxIdx = select.getSelectList().size();
+    }
+
+
+    ExtendedAliasExpander(SqlValidatorImpl validator, SqlValidatorScope scope,
+        SqlSelect select, Integer maxIdx) {
+      // This constructor is used when attempting to expand a select item, since we need to check
+      // a subset of the select list for alias information
+      // Note that it is also called recursivley, in order to handle expanding chains of aliases,
+      // IE; A as x, x as y, y as z ... etc.
+      super(validator, scope);
+      this.aliasSourceSelect = select;
+      this.maxIdx = maxIdx;
+      assert maxIdx <= select.getSelectList().size() && maxIdx >= 0;
+    }
+
+
+
+    @Override public @Nullable SqlNode visit(SqlIdentifier id) {
+      if (id.isSimple()) {
```

Comment:
Why do we require the id to be simple? This seems like it needs a comment.

Reply:
Is it a limitation in the implementation or a fundamental reason?

Reply:
I assume this is because an alias is always a simple name and if we can't resolve an id that is in the same projection it must be an alias?

Reply:
correct. I'll add a comment to this effect.